components/nimbus/src/stateful/enrollment.rs

```diff
@@ -180,6 +180,47 @@ pub fn unenroll_for_pref(
     Ok(events)
 }
 
+pub fn add_previous_state_for_experiment(
+    db: &Database,
+    writer: &mut Writer,
+    experiment_slug: &str,
+    previous_state: PreviousState,
+) -> Result<()> {
+    let enr_store = db.get_store(StoreId::Enrollments);
+
+    if let Ok(Some(existing_enrollment)) =
+        enr_store.get::<ExperimentEnrollment, Writer>(writer, experiment_slug)
+    {
+        // Previous state is only valid on Enrolled experiments
+        let updated_state = &existing_enrollment.on_add_state(previous_state);
+        enr_store.put(writer, experiment_slug, updated_state)?;
+    }
+    Ok(())
+}
+
+pub fn get_previous_state_for_experiment<'r>(
+    db: &Database,
+    reader: &'r impl Readable<'r>,
+    experiment_slug: &str,
+) -> Result<Option<PreviousState>> {
+    let enr_store = db.get_store(StoreId::Enrollments);
+
+    match enr_store.get::<ExperimentEnrollment, _>(reader, experiment_slug)? {
+        Some(enrollment) => {
+            if let EnrollmentStatus::Enrolled {
+                previous_state: Some(previous_state),
+                ..
+            } = enrollment.status
+            {
+                Ok(Some(previous_state))
+            } else {
+                Ok(None)
+            }
+        }
+        None => Ok(None),
+    }
```

Comment:
If we have an Option<T> and we have `fn: impl Fn(T) -> Option<U>` we can use `opt.and_then(fn)` [docs](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.and_then)

Additionally, the only thing that can fail here is the .get() -- everything inside the match returns `Ok(...)` so we can refactor that out.

Combining these:


```suggestion
    Ok(
        enrollments.get::<ExperimentEnrollment, _>(reader, experiment_slug)?
            .and_then(|enrollment| {
                if let EnrollmentStatus::Enrolled {
                    previous_state: Some(previous_state),
                    ..
                } = enrollment.status {
                    Some(previous_state)
                } else {
                    None
                }
            })
    )
```

Reply:
Thanks for the great explanations!